### PR TITLE
Implement `Selector` (#574 part6, closes #612)

### LINKED ISSUE #612: Let's rename elementTimeout and elementAvailabilityTimeout to selectorTimeout
\cc @VasilyStrelyaev @helen-dikareva @AlexanderMoskovkin

### PR DESCRIPTION
**Pull request by inikulin**
In this part:
- Removed `<code>` around wrapped errors in templates. They looked bad (I checked)
- Now actions and selector ctors accept selector functions, selector Promise and node snapshots as parameters
- `elementAvailabilityTimeout` -> `selectorTimeout`

It's still missing CLI text for `selector-timeout` options, because @VasilyStrelyaev decided to go to party and get wild :tada: 

\cc @helen-dikareva @AlexanderMoskovkin

### CODE REVIEW

**Comment by inikulin** [comment]
@testcafe-build-bot \retest

**Comment by inikulin** [comment]
@AlexanderMoskovkin note strange test fail in https://github.com/DevExpress/testcafe/pull/616#issuecomment-228100845

**Comment by inikulin** [comment]
@testcafe-build-bot \retest

**Comment by AlexanderMoskovkin** [comment]
> note strange test fail in #616 (comment)

I see. But I don't know what is it for now.

**Comment by AlexanderMoskovkin** [comment]
lgtm

**Comment by VasilyStrelyaev** [comment]
> It's still missing CLI text for selector-timeout options

`set the amount of time within which selector functions make attempts to obtain the page element to be returned`

Does it sound too sophisticated, or are you okay with it?

**Comment by inikulin** [comment]
> Does it sound too sophisticated, or are you okay with it?

Nah, sounds just fine, But I'd like to remove `functions` from the sentence

**Comment by helen-dikareva** [comment]
lgtm

**Comment by inikulin** [comment]
Changed CLI text, @VasilyStrelyaev pls, review and merge

File: src/cli/argument-parser.js [hunk 1]
```diff
@@ -153,7 +153,7 @@ export default class CLIArgumentParser {
             .option('-T, --test-grep <pattern>', 'run only tests matching the specified pattern')
             .option('-f, --fixture <name>', 'run only fixtures with the specified name')
             .option('-F, --fixture-grep <pattern>', 'run only fixtures matching the specified pattern')
-            .option('--element-timeout <ms>', 'set the amount of time allowed for a page element to become available before the test fails')
+            .option('--selector-timeout <ms>', 'set the amount of time within which selectors make attempts to obtain the node to be returned')
```
Comment (VasilyStrelyaev) [inline src/cli/argument-parser.js]:
```
            .option('--selector-timeout <ms>', 'set the amount of time within which selectors make attempts to obtain a node to be returned')
```

**Comment by VasilyStrelyaev** [comment]
\r-

**Comment by inikulin** [comment]
FPR

**Comment by VasilyStrelyaev** [comment]
oh, I didn't even review the rest. so one moment

File: src/client-functions/selector-factory.js [hunk 2]
```diff
@@ -2,36 +2,88 @@ import { isFinite } from 'lodash';
...
+    }
+
+    _executeCommand (args, testRun, callsite, options) {
+        var lazyPromise   = Promise.resolve();
+        var resultPromise = null;
+
+        // OPTIMIZATION: Selectors are executed lazily once they have result subscribers.
+        // It's especially useful for situations then selector result is passed to
+        // action, e.g.: `t.click(someSelector(42));` In that case we create new selector
+        // inside action, but we are not interested in already initiated execution result.
```
Comment (VasilyStrelyaev) [inline src/client-functions/selector-factory.js]:
```
        // OPTIMIZATION: Selectors are executed lazily (once someone subscribes to their result).
        // It's especially useful for situations when the selector's result is passed to
        // an action, e.g.: `t.click(someSelector(42));` In that case, we create a new selector
        // inside this action, but we are not interested in the result of ongoing execution.
```

**Comment by VasilyStrelyaev** [comment]
\r-

**Comment by inikulin** [comment]
FPR

**Comment by VasilyStrelyaev** [comment]
lgtm

**Pull request merged**